zio-config/src/main/scala/zio/config/ConfigSource.scala

```diff
@@ -2,12 +2,76 @@ package zio.config
 
 import zio.IO
 
-trait ConfigSource {
+trait ConfigSource { self =>
   val configService: ConfigSource.Service
+
+  final def orElse(that: ConfigSource): ConfigSource =
+    new ConfigSource {
+      val configService: ConfigSource.Service =
+        new ConfigSource.Service {
+          val sourceDescription: String = {
+            val selfDesc = self.configService.sourceDescription
+            val thatDesc = that.configService.sourceDescription
+
+            s"${selfDesc}: with fallback source: ${thatDesc} (on all errors)"
+          }
+
+          def getConfigValue(path: String): IO[ReadError, ConfigSource.Value] =
+            self.configService
+              .getConfigValue(path)
+              .orElse(
+                that.configService
+                  .getConfigValue(path)
+              )
+        }
+    }
+
+  final def <>(that: ConfigSource): ConfigSource =
+    orElse(that)
+
+  final def catchSome(
+    pf: PartialFunction[ReadError, ConfigSource],
+    fallbackSource: String,
+    fallbackDescription: String
+  ): ConfigSource =
+    new ConfigSource {
+      val configService: ConfigSource.Service =
+        new ConfigSource.Service {
+          val sourceDescription: String = {
+            val selfDesc = self.configService.sourceDescription
+
+            s"${selfDesc}: with fallback source: ${fallbackSource} (${fallbackDescription})"
+          }
+
+          def getConfigValue(path: String): IO[ReadError, ConfigSource.Value] =
+            self.configService
+              .getConfigValue(path)
+              .catchSome {
+                pf.andThen(
+                  _.configService
+                    .getConfigValue(path)
+                )
+              }
+        }
+    }
+
+  final def ifMissingValue(that: => ConfigSource): ConfigSource =
```

Comment:
How does this differ from `orElse`?

Also it should probably be named `orElseXYZ`

Reply:
with the current implementation, this method will take the config `that` only for a specific Error (MissingValue)
would it make sense if we choose `that` when the current config fails and we could rename it `orElse`
@Michaelt293 wdyt?

Reply:
Ah well we arlready have `orElse` ...

Reply:
Yes, that's the reason. How about the name `orElseMissing`? Not really sure if that is clear enough. More informative names could get too long, e.g.,  `orElseIfMissingValue`.